elements/lisk-transaction-pool/src/transaction_list.ts

```diff
@@ -0,0 +1,249 @@
+		if (this._nonceHeap.count > this._maxSize) {
+			this.remove(highestNonce);
+		}
+		this.events.emit(EVENT_TRANSACTION_REMOVED, {
+			address: this.address,
+			id: incomingTx.id,
+		});
+
+		return true;
+	}
+
+	public remove(nonce: bigint): boolean {
+		const removingTx = this._transactions[nonce.toString()];
+		if (!removingTx) {
+			return false;
+		}
+		// tslint:disable-next-line no-dynamic-delete
+		delete this._transactions[nonce.toString()];
+		// Recreate heap: it could remove in the middle of the heap
+		const keys = this._nonceHeap.keys;
+		this._nonceHeap.clear();
+		for (const key of keys) {
+			if (key !== nonce) {
+				this._nonceHeap.push(key, undefined);
+			}
+		}
+		this._demoteAfter(nonce);
+
+		this.events.emit(EVENT_TRANSACTION_REMOVED, {
+			address: this.address,
+			id: removingTx.id,
+		});
+
+		return true;
+	}
+
+	public promote(txs: ReadonlyArray<Transaction>): boolean {
+		// Promote if only all ID are still existing and the same
+		const promotingNonces = [];
+		for (const tx of txs) {
+			const promotingTx = this._transactions[tx.nonce.toString()];
+			if (!promotingTx) {
+				return false;
+			}
+			if (tx.id !== promotingTx.id) {
+				return false;
+			}
+			promotingNonces.push(tx.nonce);
+		}
+		this._processable = Array.from(
+			new Set([...this._processable, ...promotingNonces]),
+		).sort();
+
+		return true;
+	}
+
+	public get size(): number {
+		return this._nonceHeap.count;
+	}
+
+	public getProcessable(): ReadonlyArray<Transaction> {
+		const txs = [];
+		for (const nonce of this._processable) {
+			txs.push(this._transactions[nonce.toString()]);
+		}
+
+		return txs;
+	}
+
+	public getUnprocessable(): ReadonlyArray<Transaction> {
+		if (this._nonceHeap.count === 0) {
+			return [];
+		}
+		if (this._processable.length === this._nonceHeap.count) {
+			return [];
+		}
+		const clonedHeap = this._nonceHeap.clone();
+		// Make cloned heap root to unprocessable
+		for (const _ of this._processable) {
+			clonedHeap.pop();
+		}
+		const remainingCount = clonedHeap.count;
+		const unprocessableTx: Transaction[] = [];
+		// tslint:disable-next-line no-let
+		for (let i = 0; i < remainingCount; i += 1) {
+			const { key } = clonedHeap.pop() as { key: bigint };
+			unprocessableTx.push(this._transactions[key.toString()]);
+		}
+
+		return unprocessableTx;
+	}
+
+	public getPromotable(): ReadonlyArray<Transaction> {
+		if (this._nonceHeap.count === 0) {
+			return [];
+		}
+		if (this._processable.length === this._nonceHeap.count) {
+			return [];
+		}
+		const clonedHeap = this._nonceHeap.clone();
+		// Make cloned heap root to unprocessable
+		for (const _ of this._processable) {
+			clonedHeap.pop();
+		}
+		const firstNonProcessable = clonedHeap.pop();
+		if (!firstNonProcessable) {
+			return [];
+		}
+		if (this._processable.length !== 0) {
+			const heighestNonce = this._processable[this._processable.length - 1];
+			if (firstNonProcessable.key !== heighestNonce + BigInt(1)) {
```

Comment:
this is checking, if there is processable transaction, get the highestProcessableNonce, and the nonProcessableTransaction nonce has to be highestProcessableNonce + 1, otherwise all the nonProcessableTransaction is not possible to promote